kvserver: fix double span Finish on reproposals

pav-kv opened the pull request:
Since #106750, `ProposalData` is being copied on superseding reproposals. The caller only knows about the original proposal, and only detaches its context / tracing span when abandoning the request (e.g. on timeout). By the time of abandoning the request, there might have been a few superseding reproposals, and the context / tracing span is being used by multiple `ProposalData` objects.

In addition to rewriting the original `proposal.ctx`, we should do the same for all the `ProposalData` objects corresponding to the same request. This commit unbinds the context for older proposals when reproposing them, and unbinds the latest reproposal context when cleaning up / abandoning the request.

Fixes #107521
Fixes #108534
Fixes #108241
Fixes #108663
Fixes #108696
Fixes #108837

Epic: CRDB-25287
Release note: none

pav-kv committed:
kvserver: squash deferred statements into one

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108775)
<!-- Reviewable:end -->

pav-kv committed:
kvserver: fix double span Finish on reproposals

Since #106750, ProposalData is being copied on superseding reproposals. The
caller only knows about the original proposal, and only detaches its context /
tracing span when abandoning the request (e.g. on timeout). By the time of
abandoning the request, there might have been a few superseding reproposals,
and the context / tracing span is being used by multiple ProposalData objects.

In addition to rewriting the original proposal.ctx, we should do the same for
all the ProposalData objects corresponding to the same request. This commit
unbinds the context for older proposals when reproposing them, and unbinds the
latest reproposal context when cleaning up / abandoning the request.

Epic: CRDB-25287
Release note: none

pavelkalinnikov commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -366,17 +343,30 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
+		// context will become unused once the application of this (soft-failed)
+		// proposal concludes, i.e. soon after this method returns, in case there is
+		// anything left to log into it.
+		origP.sp = nil
+		origP.quotaAlloc = nil
+		origP.ec = makeEmptyEndCmds()
+		origP.doneCh = nil
```
Alternatively, we could put all these (maybe a few other) fields, and the context, to a struct, and reference the whole struct from `ProposalData` instead of moving/copying things between proposals.

erikgrinaker replied:
I think it's fine to keep this as it is.

erikgrinaker commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -366,17 +343,30 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
+		// proposals for post-processing.
+		//
+		// TODO(pavelkalinnikov): prove and assert that origP.reproposal was nil.
+		origP.reproposal = newProposal
```
This will prevent garbage collection of all the intermediate reproposals. I don't think we hold on to them from anywhere else? Seems like it would be worthwhile to avoid.

I'm guessing the old proposal is discarded after we hit this code path? If so, we really only need the first and last proposal in the chain. We could e.g. have a backreference to the original proposal and update its forward reference to the latest reproposal, or we could have a cancel function that's propagated down to the active reproposal.

pavelkalinnikov replied:
Reworked this bit. It became a bit more ugly, but the GC benefit seems worth it. Longer term, we should improve the lifecycle.

pavelkalinnikov replied:
Also, had to drop the invariant check.

erikgrinaker reviewed (commented):
LGTM in principle, just some minor tweaks.

pav-kv committed:
kvserv: group "moved" fields together

Epic: none
Release note: none

erikgrinaker reviewed (commented):
LGTM, but let's squash the chain/seed commits. We should also have a basic test for the proposal linking, if we can easily add one.

pavelkalinnikov commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -327,7 +327,17 @@ func (r *Replica) evalAndPropose(
+		// context to a Finished span.
+		//
+		// See https://github.com/cockroachdb/cockroach/issues/107521
+		for p := proposal; p != nil; p = p.reproposal {
```
Done.

pav-kv reviewed (commented):
Addressed inline comments, and squashed commits.

pavelkalinnikov commented:
@erikgrinaker Ah hold on, I'm looking if I can add a test.

erikgrinaker commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -379,6 +385,27 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
+		// lifecycle is reconsidered.
+		if last := seedP.lastReproposal; last != nil {
+			last.ctx = r.AnnotateCtx(context.TODO())
+		}
```
Do we expect these abandoned proposals to ever be used, or are they discarded after we return here?

I think this code makes sense as defense-in-depth, but in principle we should be able to just leave it, right?

pavelkalinnikov replied:
Logically you're right - older proposals won't be applied so we can forget them. Physically though, I'm not sure that there isn't some kind of race in result of which an older proposal comes up and "used" (for some clean-up) even after a newer proposal has been submitted.

So yeah, this is defence in depth. Need to mark it as such in the comment, and revisit.

pavelkalinnikov replied:
I would really prefer having one context/trace shared by all (re-)proposals explicitly.

pavelkalinnikov replied:
Noted in the comment.

pav-kv committed:
kvserver: add test for reproposal chaining

Epic: none
Release note: none

erikgrinaker reviewed (approved):
LGTM, thanks. It's possible that we prematurely clear the seed proposal's context here, which could possibly lose some trace events for the remainder of the failed application, but even if we do we likely don't care about them at this point anyway

pavelkalinnikov commented:
bors r=erikgrinaker

pavelkalinnikov commented:
We don't need to backport this, this refactoring is post-23.1.
